moveit_ros/trajectory_cache/src/trajectory_cache.hpp

```diff
@@ -0,0 +1,261 @@
+ *   !!! This cache does NOT support collision detection!
+ *   Trajectories will be put into and fetched from the cache IGNORING
+ *   collision.
+ *
+ *   If your planning scene is expected to change between cache lookups, do NOT
+ *   use this cache, fetched trajectories are likely to result in collision
+ *   then.
+ *
+ *   To handle collisions this class will need to hash the planning scene world
+ *   msg (after zeroing out std_msgs/Header timestamps and sequences) and do an
+ *   appropriate lookup, or do more complicated checks to see if the scene world
+ *   is "close enough" or is a less obstructed version of the scene in the cache
+ *   entry.
+ *
+ *   !!! This cache does NOT support keying on joint velocities and efforts.
+ *   The cache only keys on joint positions.
+ *
+ *   !!! This cache does NOT support multi-DOF joints.
+
+ *   !!! This cache does NOT support certain constraints
+ *   Including: path, constraint regions, everything related to collision.
+ *
+ *   This is because they are difficult (but not impossible) to implement key
+ *   logic for.
+ *
+ * Relevant ROS Parameters:
+ *   - `warehouse_plugin`: What database to use
+ *   - `warehouse_host`: Where the database should be created
+ *   - `warehouse_port`: The port used for the database
+ *
+ * This class supports trajectories planned from move_group MotionPlanRequests
+ * as well as GetCartesianPath requests. That is, both normal motion plans and
+ * cartesian plans are supported.
+ *
+ * Motion plan trajectories are stored in the `move_group_trajectory_cache`
+ * database within the database file, with trajectories for each move group
+ * stored in a collection named after the relevant move group's name.
+ *
+ * For example, the "my_move_group" move group will have its cache stored in
+ * `move_group_trajectory_cache@my_move_group`
+ *
+ * Motion Plan Trajectories are keyed on:
+ *   - Plan Start: robot joint state
+ *   - Plan Goal (either of):
+ *     - Final pose (wrt. `planning_frame` (usually `base_link`))
+ *     - Final robot joint states
+ *   - Plan Constraints (but not collision)
+ *
+ * Trajectories may be looked up with some tolerance at call time.
+ *
+ * Similarly, the cartesian trajectories are stored in the
+ * `move_group_cartesian_trajectory_cache` database within the database file,
+ * with trajectories for each move group stored in a collection named after the
+ * relevant move group's name.
+ *
+ * Cartesian Trajectories are keyed on:
+ *   - Plan Start: robot joint state
+ *   - Plan Goal:
+ *     - Pose waypoints
+ */
+class TrajectoryCache
+{
+public:
+  // We explicitly need a Node::SharedPtr because warehouse_ros ONLY supports
+  // it...
+  //
+  // TODO: methylDragon -
+  // Use rclcpp::node_interfaces::NodeInterfaces<> once warehouse_ros does.
+  explicit TrajectoryCache(const rclcpp::Node::SharedPtr& node);
+
+  bool init(const std::string& db_path = ":memory:", uint32_t db_port = 0, double exact_match_precision = 1e-6);
```

Comment:
We try to keep the documentation consistent by using the doxgen documentation style. Do you mind adding this for all the functions and classing you're introducing?

```cpp
/** \brief Once sentence to explain what the function does
 * \param [in] ParamName What does the param do
 * ...
 * \return What is the returned value (only necessary for non void return values)
 */
```

Reply:
49db82677ca60f4931979ee6deb62fa64990b84b